### ISSUE TITLE
Process resources not released on SIGKILL

### ISSUE THREAD

**Issue by marcosimioni** [shadow/shadow#1490]
**Describe the issue**

I’m trying to run a simulation where I need to start, stop, and start again the Tor process. This is a snippet of what the Shadow config looks like (the rest is just identical to the tor-minimal config shipped with Shadow 2.x):

```
  hiddenserver:
    processes:
    - path: ~/.local/bin/tgen
      args: ../../../conf/tgen.hiddenserver.graphml.xml
      start_time: 1
    - path: ~/.local/bin/tor
      args: --Address hiddenserver --Nickname hiddenserver
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      stop_time: 360
    - path: ~/.local/bin/tor
      args: --Address hiddenserver --Nickname hiddenserver
            --defaults-torrc torrc-defaults -f torrc
      start_time: 600
      stop_time: 900
```

What happens is that the first Tor process 1001 is killed (and looking at the log files I can confirm it was killed) with a SIGKILL when stop_time is reached but apparently sockets (and probably other resources?) are not properly released so the Tor process 1002 can’t start. Here is 1002’s stdout:

```
Jan 01 00:10:00.000 [notice] Tor 0.4.5.8 (git-27e5e33d2ca35de8) running on shadowsys with Libevent 2.1.8-stable, OpenSSL 1.1.1, Zlib 1.2.11, Liblzma N/A, Libzstd N/A and Glibc 2.27 as libc.
Jan 01 00:10:00.000 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
Jan 01 00:10:00.000 [notice] Read configuration file "/home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/torrc-defaults".
Jan 01 00:10:00.000 [notice] Read configuration file "/home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.common.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.common.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.client.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.client.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.hiddenserver.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.hiddenserver.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.common.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.common.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.client.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.client.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.hiddenserver.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.hiddenserver.torrc".
Jan 01 00:10:00.000 [warn] Path for DataDirectory (.) is relative and will resolve to /home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/.. Is this what you wanted?
Jan 01 00:10:00.000 [warn] Path for ServerDNSResolvConfFile (../../../conf/shadowresolv.conf) is relative and will resolve to /home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/../../../conf/s$
Jan 01 00:10:00.000 [warn] Path for HiddenServiceDir (hs) is relative and will resolve to /home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/hs. Is this what you wanted?
Jan 01 00:10:00.000 [warn] ControlPort is open, but no authentication method has been configured.  This means that any program on your computer can reconfigure your Tor.  That's bad!  You should upgrade your Tor control$
Jan 01 00:10:00.000 [warn] You have used DirAuthority or AlternateDirAuthority to specify alternate directory authorities in your configuration. This is potentially dangerous: it can make you look different from all oth$
Jan 01 00:10:00.000 [warn] TestingTorNetwork is set. This will make your node almost unusable in the public Tor network, and is therefore only advised if you are building a testing Tor network!
Jan 01 00:10:00.000 [notice] Opening Socks listener on 127.0.0.1:9000
Jan 01 00:10:00.000 [warn] Could not bind to 127.0.0.1:9000: Address already in use. Is Tor already running?
Jan 01 00:10:00.000 [notice] Opening Control listener on 127.0.0.1:9051
Jan 01 00:10:00.000 [warn] Could not bind to 127.0.0.1:9051: Address already in use. Is Tor already running?
Jan 01 00:10:00.000 [warn] Failed to parse/validate config: Failed to bind one of the listener ports.
Jan 01 00:10:00.000 [err] Reading config failed--see warnings above.
```

I’ve tried then to re-run the simulation and manually terminate the first Tor process with a SIGTERM (`kill <pid>`) and the Tor process 1001 exits cleanly:

```
$ tail hiddenserver.tor.1001.stdout -n 20
Jan 01 00:03:04.703 [info] pathbias_count_build_success(): Got success count 11.000000/11.000000 for guard relay1 ($3FB0BD7827C760FE7F9DD810FCB10322D63AB4CF)
Jan 01 00:03:04.703 [info] hs_circ_service_intro_has_opened(): Introduction circuit just opened but we have enough for service r4aj4kaqf46mala2yykldkvwrrwjagab2qppuqtvgdxwh6spsulwu2qd. Repurposing it to general and leaving internal.
Jan 01 00:03:04.703 [notice] Catching signal TERM, exiting cleanly.
Jan 01 00:03:04.703 [info] or_state_save(): Saved state to "./state"
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 1) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 15) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 14) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 13) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 12) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 11) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 10) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 9) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 8) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 7) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 6) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 5) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 4) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 3) has been freed.
Jan 01 00:03:04.703 [info] circuit_free_(): Circuit 0 (id: 2) has been freed.
Jan 01 00:03:04.703 [info] tor_lockfile_unlock(): Unlocking "./lock"
```

And the subsequent new Tor process 1002 runs fine:

```
$ head hiddenserver.tor.1002.stdout -n 40
Jan 01 00:10:00.000 [notice] Tor 0.4.5.8 (git-27e5e33d2ca35de8) running on shadowsys with Libevent 2.1.8-stable, OpenSSL 1.1.1, Zlib 1.2.11, Liblzma N/A, Libzstd N/A and Glibc 2.27 as libc.
Jan 01 00:10:00.000 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
Jan 01 00:10:00.000 [notice] Read configuration file "/home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/torrc-defaults".
Jan 01 00:10:00.000 [notice] Read configuration file "/home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.common.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.common.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.client.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.client.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.hiddenserver.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.hiddenserver.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.common.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.common.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.client.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.client.torrc".
Jan 01 00:10:00.000 [notice] Processing configuration path "../../../conf/tor.hiddenserver.torrc" at recursion level 1.
Jan 01 00:10:00.000 [notice] Including configuration file "../../../conf/tor.hiddenserver.torrc".
Jan 01 00:10:00.000 [warn] Path for DataDirectory (.) is relative and will resolve to /home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/.. Is this what you wanted?
Jan 01 00:10:00.000 [warn] Path for ServerDNSResolvConfFile (../../../conf/shadowresolv.conf) is relative and will resolve to /home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/../../../conf/shadowresolv.conf. Is this what you wanted?
Jan 01 00:10:00.000 [warn] Path for HiddenServiceDir (hs) is relative and will resolve to /home/*****/tor/shadow/src/test/tor/minimal/shadow.data/hosts/hiddenserver/hs. Is this what you wanted?
Jan 01 00:10:00.000 [warn] ControlPort is open, but no authentication method has been configured.  This means that any program on your computer can reconfigure your Tor.  That's bad!  You should upgrade your Tor controller as soon as possible.
Jan 01 00:10:00.000 [warn] You have used DirAuthority or AlternateDirAuthority to specify alternate directory authorities in your configuration. This is potentially dangerous: it can make you look different from all other Tor users, and hurt your anonymity. Even if you've specified the same authorities as Tor uses by default, the defaults could change in the future. Be sure you know what you're doing.
Jan 01 00:10:00.000 [warn] TestingTorNetwork is set. This will make your node almost unusable in the public Tor network, and is therefore only advised if you are building a testing Tor network!
Jan 01 00:10:00.000 [notice] Opening Socks listener on 127.0.0.1:9000
Jan 01 00:10:00.000 [notice] Opened Socks listener connection (ready) on 127.0.0.1:9000
Jan 01 00:10:00.000 [notice] Opening Control listener on 127.0.0.1:9051
Jan 01 00:10:00.000 [notice] Opened Control listener connection (ready) on 127.0.0.1:9051
Jan 01 00:10:00.000 [warn] Your log may contain sensitive information - you disabled SafeLogging, and you're logging more than "notice". Don't log unless it serves an important reason. Overwrite the log afterwards.
Jan 01 00:10:00.000 [info] options_commit_listener_transaction(): Recomputed OOS thresholds: ConnLimit 1000, ConnLimit_ 1048544, ConnLimit_high_thresh 1048480, ConnLimit_low_thresh 786408
Jan 01 00:10:00.000 [info] crypto_openssl_late_init(): NOT using OpenSSL engine support.
Jan 01 00:10:00.000 [info] evaluate_evp_for_aes(): This version of OpenSSL has a known-good EVP counter-mode implementation. Using it.
Jan 01 00:10:00.000 [notice] Not disabling debugger attaching for unprivileged users.
Jan 01 00:10:00.000 [info] tor_lockfile_lock(): Locking "./lock"
Jan 01 00:10:00.000 [info] config_generic_service(): HiddenServiceDir="hs". Configuring...
Jan 01 00:10:00.000 [info] config_generic_service(): HiddenServicePort=80 127.0.0.1:8080 for "hs"
Jan 01 00:10:00.000 [info] config_service_v3(): Service INTRO2 DoS defenses rate set to: 25
Jan 01 00:10:00.000 [info] config_service_v3(): Service INTRO2 DoS defenses burst set to: 200
Jan 01 00:10:00.000 [info] or_state_load(): Loaded state from "./state"
Jan 01 00:10:00.000 [info] sampled_guards_update_from_consensus(): Not updating the sample guard set; we have no reasonably live consensus.
Jan 01 00:10:00.000 [info] first_reachable_filtered_entry_guard(): Trying to sample a reachable guard: We know of 0 in the USABLE_FILTERED set.
Jan 01 00:10:00.000 [info] first_reachable_filtered_entry_guard():   (That isn't enough. Trying to expand the sample.)
```

Why do you think sockets are not being released on SIGKILL? Is this the intended behavior?

**To Reproduce**
Please modify the hiddenservice host of the tor-minimal example as reported above.

**Operating System (please complete the following information):**
 - OS and version: post the output of `lsb_release -d`  
```
$ lsb_release -d
Description:    Ubuntu 18.04.5 LTS
```
 - Kernel version: post the output of `uname -a`  
```
$ uname -a
Linux ***** 4.15.0-137-generic #141-Ubuntu SMP Fri Feb 19 13:46:27 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Shadow (please complete the following information):**
 - Version: post the output of `shadow --version`  
```
$ shadow --version
Shadow 2.0.0-pre.1
```
 - Which plug-ins you are using:  
Tor

**Additional context**

**Comment by stevenengler** [comment]
> I think what I’m describing below are two separate issues - an issue with process resources not being correctly released when Shadow kills a process with a SIGKILL, and the inability for Shadow to terminate a process with SIGTERM.
>
> I’m leaving them together for now until I hear your suggestions.

Thanks for the bug report! I think the second issue should be moved to a new issue. The first one definitely looks like a bug, and the second one is more of a feature request.

**Comment by marcosimioni** [comment]
@stevenengler done! https://github.com/shadow/shadow/issues/1491

however please note I don't have the ability to create a "feature request" type of issue (I only have the option to create a bug, or a discussion), so you may want to re-label that.

**Comment by stevenengler** [comment]
This issue should be fixed in #1493. The issue was that shadow wasn't disassociating the sockets from the network interface when the process was stopped. It worked correctly when you used a SIGTERM since tor has a signal handler that will close the sockets manually when it shuts down. Your config file should work now, but with an extra step to cleanup the lock file since we use SIGKILL to stop the process:

```yaml
  hiddenserver:
    processes:
    - path: ~/.local/bin/tgen
      args: ../../../conf/tgen.hiddenserver.graphml.xml
      start_time: 1
    - path: ~/.local/bin/tor
      args: --Address hiddenserver --Nickname hiddenserver
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      stop_time: 360
    # /usr/bin/rm doesn't seem to work in shadow, so use python instead
    - path: /usr/bin/python3
      args: -c "import os; os.remove('lock')"
      start_time: 361
    - path: ~/.local/bin/tor
      args: --Address hiddenserver --Nickname hiddenserver
            --defaults-torrc torrc-defaults -f torrc
      start_time: 600
      stop_time: 900
```

**Comment by marcosimioni** [comment]
Thank you! Just for the records and in case this comes useful to others, I temporarily wrote a small application as a workaround that I named stop_tor in order to gracefully shut-down Tor via control port:

```
#include <stdio.h> 
#include <stdlib.h> 
#include <errno.h> 
#include <string.h> 
#include <netdb.h> 
#include <sys/types.h> 
#include <netinet/in.h> 
#include <sys/socket.h> 
#include <unistd.h>

#define PORT 9051
#define MAXDATASIZE 100

int main(int argc, char *argv[])
{
    int sockfd, numbytes;  
    char buf[MAXDATASIZE];
    struct hostent *he;
    struct sockaddr_in their_addr;

    if ((he=gethostbyname("localhost")) == NULL) { 
        herror("gethostbyname");
        exit(1);
    }

    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
        perror("socket");
        exit(1);
    }

    their_addr.sin_family = AF_INET;      /* host byte order */
    their_addr.sin_port = htons(PORT);    /* short, network byte order */
    their_addr.sin_addr = *((struct in_addr *)he->h_addr);
    bzero(&(their_addr.sin_zero), 8);     /* zero the rest of the struct */

    if (connect(sockfd, (struct sockaddr *)&their_addr, \
        sizeof(struct sockaddr)) == -1) {
        perror("connect");
        exit(1);
    }

    const char *msg1 = "authenticate \"\"\n";
    if (send(sockfd, msg1, strlen(msg1), 0) == -1){
        perror("send");
        exit (1);
    }
    printf("Msg sent: %s\n", msg1);
    sleep(1);

    if ((numbytes=recv(sockfd, buf, MAXDATASIZE, 0)) == -1) {
        perror("recv");
        exit(1);
    }
    buf[numbytes] = '\0';
    printf("Received in pid=%d, text=: %s \n",getpid(), buf);
    sleep(1);

    const char *msg2 = "SIGNAL SHUTDOWN\n";
    if (send(sockfd, msg2, strlen(msg2), 0) == -1){
        perror("send");
        exit (1);
    }
    printf("Msg sent: %s\n", msg2);
    sleep(1);

    if ((numbytes=recv(sockfd, buf, MAXDATASIZE, 0)) == -1) {
        perror("recv");
        exit(1);
    }
    buf[numbytes] = '\0';
    printf("Received in pid=%d, text=: %s \n",getpid(), buf);
    sleep(1);

    close(sockfd);

    return 0;
}
```

In the shadow.config, you just have to make sure Tor listens on the control port, and then you can have this small app triggering whenever you want to shut Tor down as follows:

```
  hiddenservice1:
    options:
      country_code_hint: DE
      pcap_directory: pcap
    bandwidth_up: 100000000 kilobit
    bandwidth_down: 100000000 kilobit
    processes:
    - path: ~/.local/bin/tgen
      args: ../../../conf/tgen-server.tgenrc.graphml
      start_time: 300
    - path: ~/.local/bin/tor
      args: --Address hiddenservice1 --Nickname hiddenservice1 --defaults-torrc torrc-defaults --controlport 9051
        -f torrc
      start_time: 240
    - path: ~/.local/bin/stop_tor
      start_time: 1500
    - path: ~/.local/bin/tor
      args: --Address hiddenservice1 --Nickname hiddenservice1 --defaults-torrc torrc-defaults --controlport 9051
        -f torrc
      start_time: 1800
    - path: ~/.local/bin/stop_tor
      start_time: 3300
```

Thank you for fixing the issue with SIGKILL, I will probably revert to that.
